Unable to find first hop -> crash

This only happens sometimes, and the reasons for the first hop lookup failing aren't clear, but whenever it does fail, the error isn't handled correctly, and leads to an assertion failure soon after. 

https://circleci.com/gh/netsec-ethz/scion/3832 is an example.
```
2016-10-17 12:55:57.852771+00:00 [ERROR] (MainThread) Unable to find first hop:
<SCION-Path>
  <A-Segment>
    InfoOpaqueField(8B): flags: UP, TS: 2016-10-17 12:55:52+00:00, ISD: 2, hops: 3
    HopOpaqueField(8B): flags: None, exp_time: 63, ingress: 0, egress: 35, mac: 61c638
    HopOpaqueField(8B): flags: None, exp_time: 63, ingress: 96, egress: 54, mac: 7f9681
    HopOpaqueField(8B): flags: None, exp_time: 63, ingress: 88, egress: 0, mac: 4b29fc
  </A-Segment>
</SCION-Path>
2016-10-17 12:55:58.249931+00:00 [DEBUG] (Elem.packet_recv) _get_msg_meta() called
2016-10-17 12:55:58.251552+00:00 [CRITICAL] (MainThread) Exception in main process:
2016-10-17 12:55:58.252800+00:00 [CRITICAL] (MainThread) Traceback (most recent call last):
2016-10-17 12:55:58.253085+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/lib/main.py", line 37, in main_wrapper
2016-10-17 12:55:58.253398+00:00 [CRITICAL] (MainThread)     main(*args, **kwargs)
2016-10-17 12:55:58.253663+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/lib/main.py", line 81, in main_default
2016-10-17 12:55:58.254229+00:00 [CRITICAL] (MainThread)     inst.run()
2016-10-17 12:55:58.254786+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/path_server/base.py", line 380, in run
2016-10-17 12:55:58.255248+00:00 [CRITICAL] (MainThread)     super().run()
2016-10-17 12:55:58.255512+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/scion_elem.py", line 513, in run
2016-10-17 12:55:58.255774+00:00 [CRITICAL] (MainThread)     self._packet_process()
2016-10-17 12:55:58.256034+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/scion_elem.py", line 619, in _packet_process
2016-10-17 12:55:58.256307+00:00 [CRITICAL] (MainThread)     self.handle_msg_meta(*self._in_buf.get(timeout=1.0))
2016-10-17 12:55:58.256594+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/scion_elem.py", line 234, in handle_msg_meta
2016-10-17 12:55:58.256906+00:00 [CRITICAL] (MainThread)     handler(msg, meta)
2016-10-17 12:55:58.257297+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/path_server/base.py", line 274, in handle_path_segment_record
2016-10-17 12:55:58.257806+00:00 [CRITICAL] (MainThread)     added.update(self._dispatch_segment_record(type_, pcb, **params))
2016-10-17 12:55:58.258290+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/path_server/base.py", line 285, in _dispatch_segment_record
2016-10-17 12:55:58.258767+00:00 [CRITICAL] (MainThread)     return handle_map[type_](seg, **kwargs)
2016-10-17 12:55:58.259278+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/path_server/core.py", line 137, in _handle_core_segment_record
2016-10-17 12:55:58.259715+00:00 [CRITICAL] (MainThread)     self._handle_waiting_targets(pcb, reverse=reverse)
2016-10-17 12:55:58.260151+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/path_server/base.py", line 325, in _handle_waiting_targets
2016-10-17 12:55:58.260721+00:00 [CRITICAL] (MainThread)     self._send_waiting_queries(dst_ia[0], pcb)
2016-10-17 12:55:58.261266+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/path_server/base.py", line 337, in _send_waiting_queries
2016-10-17 12:55:58.261791+00:00 [CRITICAL] (MainThread)     self.send_meta(seg_req, meta)
2016-10-17 12:55:58.262312+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/scion_elem.py", line 457, in send_meta
2016-10-17 12:55:58.262829+00:00 [CRITICAL] (MainThread)     return self.send(pkt, *next_hop_port)
2016-10-17 12:55:58.263347+00:00 [CRITICAL] (MainThread)   File "/home/scion/go/src/github.com/netsec-ethz/scion/infrastructure/scion_elem.py", line 431, in send
2016-10-17 12:55:58.263875+00:00 [CRITICAL] (MainThread)     assert isinstance(dst_port, int), dst_port
2016-10-17 12:55:58.264439+00:00 [CRITICAL] (MainThread) AssertionError
2016-10-17 12:55:58.265000+00:00 [CRITICAL] (MainThread) 
2016-10-17 12:55:58.265338+00:00 [CRITICAL] (MainThread) Exiting

```

--- a/infrastructure/path_server/base.py
+++ b/infrastructure/path_server/base.py
@@ -312,13 +312,11 @@ class PathServer(SCIONElement, metaclass=ABCMeta):
         """
         raise NotImplementedError
 
-    def _handle_waiting_targets(self, pcb, reverse=False):
+    def _handle_waiting_targets(self, pcb):
         """
         Handle any queries that are waiting for a path to any core AS in an ISD.
         """
         dst_ia = pcb.first_ia()
-        if reverse:
-            dst_ia = pcb.last_ia()
         if not self.is_core_as(dst_ia):
             logging.warning("Invalid waiting target, not a core AS: %s", dst_ia)
             return

--- a/infrastructure/path_server/base.py
+++ b/infrastructure/path_server/base.py
@@ -322,13 +322,13 @@ class PathServer(SCIONElement, metaclass=ABCMeta):
         if not self.is_core_as(dst_ia):
             logging.warning("Invalid waiting target, not a core AS: %s", dst_ia)
             return
-        self._send_waiting_queries(dst_ia[0], pcb, not reverse)
+        self._send_waiting_queries(dst_ia[0], pcb)
 
-    def _send_waiting_queries(self, dst_isd, pcb, reverse_path):
+    def _send_waiting_queries(self, dst_isd, pcb):
         targets = self.waiting_targets[dst_isd]
         if not targets:
             return
-        path = pcb.get_path(reverse_direction=reverse_path)
+        path = pcb.get_path(reverse_direction=True)
         src_ia = pcb.first_ia()
         while targets:
             seg_req = targets.pop(0)

--- a/infrastructure/path_server/base.py
+++ b/infrastructure/path_server/base.py
@@ -322,13 +322,13 @@ class PathServer(SCIONElement, metaclass=ABCMeta):
         if not self.is_core_as(dst_ia):
             logging.warning("Invalid waiting target, not a core AS: %s", dst_ia)
             return
-        self._send_waiting_queries(dst_ia[0], pcb)
+        self._send_waiting_queries(dst_ia[0], pcb, not reverse)
 
-    def _send_waiting_queries(self, dst_isd, pcb):
+    def _send_waiting_queries(self, dst_isd, pcb, reverse_path):
         targets = self.waiting_targets[dst_isd]
         if not targets:
             return
-        path = pcb.get_path(reverse_direction=True)
+        path = pcb.get_path(reverse_direction=reverse_path)
         src_ia = pcb.first_ia()
         while targets:
             seg_req = targets.pop(0)

--- a/infrastructure/path_server/core.py
+++ b/infrastructure/path_server/core.py
@@ -133,8 +133,9 @@ class CorePathServer(PathServer):
         if not added:
             return set()
         # Send pending requests that couldn't be processed due to the lack of
-        # a core segment to the destination PS.
-        self._handle_waiting_targets(pcb, reverse=reverse)
+        # a core segment to the destination PS. Don't use SIBRA PCBs for that.
+        if not reverse:
+            self._handle_waiting_targets(pcb)
         ret = set([(first_ia, pcb.is_sibra())])
         if first_ia[0] != self.addr.isd_as[0]:
             # Remote core segment, signal the entire ISD

--- a/infrastructure/path_server/core.py
+++ b/infrastructure/path_server/core.py
@@ -134,7 +134,7 @@ class CorePathServer(PathServer):
             return set()
         # Send pending requests that couldn't be processed due to the lack of
         # a core segment to the destination PS. Don't use SIBRA PCBs for that.
-        if not reverse:
+        if not pcb.is_sibra():
             self._handle_waiting_targets(pcb)
         ret = set([(first_ia, pcb.is_sibra())])
         if first_ia[0] != self.addr.isd_as[0]:

--- a/infrastructure/scion_elem.py
+++ b/infrastructure/scion_elem.py
@@ -451,7 +451,7 @@ class SCIONElement(object):
                                  meta.ia, msg, dst_port)
         if not next_hop_port:
             next_hop_port = self.get_first_hop(pkt)
-        if not next_hop_port:
+        if next_hop_port == (None, None):
             logging.error("Can't find first hop, dropping packet\n%s", pkt)
             return False
         return self.send(pkt, *next_hop_port)

